rfcs/0010-txpool-dos-mitigation.md

```diff
@@ -0,0 +1,238 @@
+resources are paid for at a price that network participants would be happy with.
+
+In this context, we have the following goals:
+
+1.  Support a pool of pending transactions with a maximum size, evicting
+    transactions when at the maximum size based on fee.
+2.  Allow multiple queued transactions from one address. Since payments aren't
+    instant, users will want this. We could make transaction senders responsible
+    for queuing, but that wouldn't allow multiple transactions sent from the
+    same address to be included in the same block.
+3.  Allow for transaction replacement, while aligning incentives by charging for
+    it appropriately. Users can e.g. cancel payments by replacing them with a
+    no-op transaction (send themselves $0) or resend them with a higher fee if
+    they're processing too slow. This is important for two reasons. A) a
+    transaction sent with too low of a fee will get "stuck", and without the
+    ability to replace it the user's account is bricked until the transaction
+    ages out of the txpool or is evicted if the pool is full. B) a transaction
+    sent with too low of a fee may simply take long enough that the user would
+    rather not have sent it. E.g. if I'm buying a coffee I don't want to wait in
+    the shop for half an hour while my payment goes through. Or if I'm using a
+    market based on smart contracts and my trade execution is delayed, the reason
+    I wanted to make the trade may no longer be valid when it actually executes.
+
+### When we first receive a transaction: rules
+
+When we receive a gossiped transaction, we will check the below constraints,
+with respect to our current longest chain. If any of the checks fail, we ignore
+the transaction and do not gossip it.
+
+1.  The signature.
+2.  The sender's account exists.
+3.  The sender's balance (inclusive of any pending txs from the same account
+    with lower nonce) is >= the transaction amount + fee.
+4.  The tx nonce is equal to the sending account's next nonce, inclusive of
+    transactions already in the pool. (If it conflicts with one in the pool, see
+    below.)
+
+If we get any transactions where the signature is invalid, we blacklist the
+sending node immediately. No honest node sends transactions like that. If we get
+transactions that fail for other reasons - the ones that depend on the current
+chain state - then we increment the node's bad tx count. We track the bad/good
+tx ratio, and blacklist the node if it exceeds `bad_tx_blacklist_ratio`. Honest
+nodes that are out of date might innocently send us transactions that are not
+valid, but they won't send us a lot of them. If a node detects it is
+substantially behind the network, it should disable gossip until it catches up.
+We give new peers `good_faith_tx_credits` good transaction credits when we first
+connect to them. That way we won't erroneously blacklist honest peers early on.
```

Comment:
we should somehow integrate this with the existing punishment system (cc @cmr )

Reply:
yeah we still need to refine the banlist policy. fortunately it should slot into the mechanism quite well.